src/app/zones/sources.ts

```diff
@@ -14,34 +16,54 @@ type DetailParams = {
   name: string
 }
 
-type Closeable = { close: () => void }
-
 export type ZoneOverviewCollection = CollectionResponse<ZoneOverview>
 export type ZoneOverviewSource = DataSourceResponse<ZoneOverview>
 export type ZoneOverviewCollectionSource = DataSourceResponse<ZoneOverviewCollection>
 
 export type EnvoyDataSource = DataSourceResponse<object | string>
 
-export const sources = (api: KumaApi) => {
+export const sources = (source: Source, api: KumaApi) => {
   return {
     ...zoneIngresses(api),
     ...zoneEgresses(api),
 
-    '/zone-cps': async (params: PaginationParams, source: Closeable) => {
-      source.close()
-
+    '/zone-cps': async (params: PaginationParams) => {
       const { size } = params
-      const offset = params.size * (params.page - 1)
-
-      return await api.getAllZoneOverviews({ size, offset })
+      const offset = size * (params.page - 1)
+      return api.getAllZoneOverviews({ size, offset })
     },
 
-    '/zone-cps/:name': async (params: DetailParams, source: Closeable) => {
-      source.close()
-
+    '/zone-cps/:name': async (params: DetailParams) => {
       const { name } = params
-
-      return await api.getZoneOverview({ name })
+      return api.getZoneOverview({ name })
+    },
+    '/zone-cps/online/:name': (params: DetailParams) => {
+      // this source retries until we have a 200 and the zone found is online
+      // any non-404 errors will error as usual
+      const { name } = params
+      return source(async () => {
+        const res = await api.getZoneOverview({ name })
+        if (getStatus(res.zoneInsight?.subscriptions) === 'online') {
+          return res
+        } else {
+          const e = new ApiError({
+            status: 404,
+            title: `The ${res.name} Zone is offline`,
+          })
+          throw e
+        }
+      }, {
+        retry: (e) => {
+          if (typeof e.status !== 'undefined') {
+            const status = e.status.toString()
+            switch (true) {
+              case status === '404':
+                // wait 2 seconds and try again
+                return new Promise((resolve) => setTimeout(resolve, 2000))
+            }
+          }
```

Comment:
**Suggestion**: With a better type on the `retry` callback, this should be as simple as that:

```suggestion
          if (e.status === 404) {
            // wait 2 seconds and try again
            return new Promise((resolve) => setTimeout(resolve, 2000))
          }
```

Also, why convert the status code to a string first only to then compare it to a string literal?

Why use a switch statement for a one-branch check that can be a simple if statement instead? It seems very unlikely this would need to be extended to any other status code.

Reply:
Unless I'm mistaken `e` can be more than just something with a `.status`?

---

I prefer generally using strings for http error code checks as quite quickly, being able to check for things like `4xx` or `5xx`, becomes helpful. In which case you can just use startsWith.

Similarly, for these types of checks it good to start as you mean to go on. `switch` would be much better here (if) it grows, so starting with `switch` means people that come here to add to this are likely to stay with the existing `switch` instead of using lots of `||`s.

Will we need to add things here? Maybe, maybe not. Is is likely someone will copy this paste it elsewhere and amend? Maybe, maybe not. But if any of those cases are a "yes", then we are starting out as we should go on, using `switch`.

Reply:
1. What's the technical problem with `switch`?
2. I almost used a different error type/error code for the "zone exists but is offline". It may even be better to do that at some point in the future as this state isn't technically 404. In which case the `if` conditional will get more complicated.

I think overall this is more of a general idea of "use strings/switch for forking via HTTP status codes". Its rather like a state machine for which `switch` is extremely common. I pointed out the "string thing (5xx)" and its usefulness before and considering there could be lots of eventualities, it could end up being a complicated conditional if you aren't careful. Lastly as this is likely to be copy pasted elsewhere, then still I would much rather start as I would like it to go on, using `switch`.

I think the above plus, the fact that we already nearly missed that this would throw more than just an ApiError makes me want to keep this as is i.e. logic based on a `.status: String` using the style of a simple state machine, the status being the `state`.

I think largely this isn't something we need to spend much time talking about or fiddling with.

Reply:
I think I can resolve this, but let me know if not.

Reply:
It’s still a premature optimization if you ask me, but so be it.